### PR TITLE
LAprop as an optimizer along with frobenius weight normalization + associated tests

### PR DESCRIPTION
**Pull request by mkhona-nvidia**
Convert the LAProp update function into a scalar optimizer

### CODE REVIEW

**Comment by skyw** [comment]
Shall we seek more programmable way to make optimizer out of scalar update function? Instead of adding them one by one (with a lot of duplicated code) when we need it?

File: emerging_optimizers/scalar_optimizers/laprop.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    def __init__(
+        self,
+        params: ParamsT,
+        lr: float = 1e-3,
+        betas: tuple[float, float] = (0.9, 0.999),
+        eps: float = 1e-8,
+        weight_decay: float = 0.0,
+        correct_bias: bool = True,
+        normalize: bool = False,
+        normalize_eps: float = 1e-8,
```
Comment (skyw) [inline emerging_optimizers/scalar_optimizers/laprop.py]: Nit: Do we need two eps? They are for different things, but I feel having two of them only increase probability to mess up one of them.

File: emerging_optimizers/scalar_optimizers/laprop.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+        Args:
+            closure: A closure that reevaluates the model and returns the loss (optional).
+
+        Returns:
+            The loss from the closure, if provided.
+        """
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
```
Comment (skyw) [inline emerging_optimizers/scalar_optimizers/laprop.py]: I know greptile like to signal this, but all of our optimizer are not supposed to have gradient in it. can just assert closure is None and ignore greptile.

File: emerging_optimizers/scalar_optimizers/laprop.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            weight_decay = group["weight_decay"]
+            correct_bias = group["correct_bias"]
+
+            for p in group["params"]:
+                if p.grad is None:
+                    continue  # pragma: no cover
+
+                grad = p.grad
+                state = self.state[p]
+                state["step"] += 1
```
[the remark this answers is not in the document]
Comment (skyw) [reply emerging_optimizers/scalar_optimizers/laprop.py:167]: Seems a valid concern?
I'd say either fix or just remove weight_decay_method in argument and only use decoupled.

File: emerging_optimizers/scalar_optimizers/laprop.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                    state["exp_avg"],
+                    state["exp_avg_sq"],
+                    correct_bias,
+                    betas,
+                    state["step"],
+                    eps,
+                )
+                pre_norm = p.data.norm() if self.normalize else None
+                p.data.add_(update, alpha=-lr)
+                if pre_norm is not None:
```
Comment (skyw) [inline emerging_optimizers/scalar_optimizers/laprop.py]: Use self.normalize as condition.

File: emerging_optimizers/scalar_optimizers/laprop.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    def __init__(
+        self,
+        params: ParamsT,
+        lr: float = 1e-3,
+        betas: tuple[float, float] = (0.9, 0.999),
+        eps: float = 1e-8,
+        weight_decay: float = 0.0,
+        correct_bias: bool = True,
+        normalize: bool = False,
```
Comment (skyw) [inline emerging_optimizers/scalar_optimizers/laprop.py]: Nit: Consider a more informative name to indicate what normalization the flag enables.

File: tests/test_laprop.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        optimizer.step()
+
+        self.assertEqual(optimizer.state[param]["step"], 1)
+        self.assertIn("exp_avg", optimizer.state[param])
+        self.assertIn("exp_avg_sq", optimizer.state[param])
+
+        expected_exp_avg_sq = (1 - beta2) * grad.square()
+        normalized_grad = grad / (grad.abs() + optimizer.param_groups[0]["eps"])
+        expected_exp_avg = (1 - beta1) * normalized_grad
+        torch.testing.assert_close(optimizer.state[param]["exp_avg_sq"], expected_exp_avg_sq)
```
Comment (skyw) [inline tests/test_laprop.py]: Q: Should they exact match?

File: tests/test_laprop.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        old_param = param.detach().clone()
+        exp_avg = torch.zeros_like(param)
+        exp_avg_sq = torch.zeros_like(param)
+        expected_update = calculate_laprop_update(grad, exp_avg, exp_avg_sq, True, betas, 1, eps)
+
+        param.grad = grad.clone()
+        optimizer.step()
+
+        torch.testing.assert_close(param, old_param - lr * expected_update)
+        torch.testing.assert_close(optimizer.state[param]["exp_avg"], exp_avg)
```
Comment (skyw) [inline tests/test_laprop.py]: Test name suggest "match", both atol and rtol needs to be 0.

**Review by skyw — CHANGES_REQUESTED** [review]

**Review by skyw — DISMISSED** [review]
We should do a full repo clean up for the closure. We don't support it in any of our optimizer, i.e. none was tested, should just error out if closure is not none.

**Review by skyw — APPROVED** [review]

**Review by skyw — APPROVED** [review]
